core/src/storage2/collections/stash/mod.rs

```diff
@@ -0,0 +1,549 @@
+                Some(value)
+            }
+            Entry::Vacant { .. } => {
+                unreachable!("the taken entry is known to be occupied")
+            }
+        }
+    }
+
+    /// Defragments the underlying storage to minimize footprint.
+    ///
+    /// Returns the number of storage cells freed this way.
+    ///
+    /// This might invalidate indices stored outside of the stash.
+    ///
+    /// # Callback
+    ///
+    /// In order to keep those indices up-to-date the caller can provide
+    /// a callback function that is called for every moved entry
+    /// with a shared reference to the entries value and the old as well
+    /// as the new index.
+    ///
+    /// # Note
+    ///
+    /// - If `max_iterations` is `Some` concrete value it is used in order to
+    ///   bound the number of iterations and won't try to defrag until the stash
+    ///   is optimally compacted.
+    /// - Users are adviced to call this method using `Some` concrete
+    ///   value to keep gas costs within certain bounds.
+    /// - The call to the given callback takes place before the reinsertion
+    ///   of the shifted occupied entry.
+    pub fn defrag<C>(&mut self, max_iterations: Option<u32>, mut callback: C) -> u32
+    where
+        C: FnMut(Index, Index, &T),
+    {
+        let len_entries = self.len_entries();
+        let mut freed_cells = 0;
+        for index in (0..len_entries)
+            .rev()
+            .take(max_iterations.unwrap_or(len_entries) as usize)
+        {
+            if !self.has_vacant_entries() {
+                // Bail out as soon as there are no more vacant entries left.
+                return freed_cells
+            }
+            // In any case we are going to free yet another storage cell.
+            freed_cells += 1;
+            match self
+                .entries
+                .put_get(index, None)
+                .expect("index is out of bounds")
+            {
+                Entry::Vacant(vacant_entry) => {
+                    // Remove the vacant entry and rebind its neighbours.
+                    self.remove_vacant_entry(index, vacant_entry);
+                }
+                Entry::Occupied(value) => {
+                    // Move the occupied entry into one of the remaining vacant
+                    // entries. We do not re-use the `put` method to not update
+                    // the length and other header information.
+                    let vacant_index = self
+                        .last_vacant_index()
+                        .expect("it has been asserted that there are vacant entries");
+                    callback(index, vacant_index, &value);
+                    let new_entry = Some(Entry::Occupied(value));
+                    let old_entry = self
+                        .entries
+                        .put_get(vacant_index, new_entry)
+                        .expect("a `next_vacant` index must point to an occupied cell");
```

Comment:
Expect message appears out of date perhaps `next_index` refers to `last_vacant_index()`

Reply:
to be honest I am not sure about those changes anymore. it has been a while since I touched this code. :sweat_smile:

Reply:
done